### DISCUSSION TITLE
Shipment transition callbacks being skipped

### DISCUSSION THREAD

**Discussion by skukx** [solidusio/solidus#4619] [New Features or Ideas]
Assume the below setup:

```ruby
Spree::Shipment.class_eval do
  state_machine.after_transition to: :ready, do: :my_method

  def my_method
    puts 'my_method called'
  end
end
```

After an order is completed, `Spree::Order#finalize!` gets called.
See https://github.com/solidusio/solidus/blob/86501e40827ed9a5eb8eccb4f2dd9a2198c1e0a1/core/app/models/spree/order.rb#L409-L412

If the order is paid for then these shipments will be updated to `ready`. The problem is that this update skips callbacks so `Spree::Shipment#my_method` will not be called even though the shipment has transitioned to a state of `ready`.

Expected behavior would be that when the order moved to `complete` and updates the shipments. Each shipment that is updated to another state will create a transition event for that state change.

Solidus 2.2
Happens in Core Solidus

See https://github.com/solidusio/solidus/pull/2065 for specs recreating behavior

**Comment by mamhoff** [comment]
This is a can of worms. We explicitly don't run the Shipment state machine, and rather determine the shipment's state by its attributes for performance and complexity reasons (the order and shipment state machine interact in strange ways). 

You're right in that we should do one of two things: Either remove the shipment state machine entirely, or actually use it.
